src/memory.rs

```diff
@@ -0,0 +1,41 @@
+use std::mem;
+use std::slice;
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Buffer {
+    pub ptr: *mut u8,
+    pub size: usize,
+}
+
+// this frees memory we released earlier
+#[no_mangle]
+pub extern "C" fn free_rust(buf: Buffer) {
+    unsafe {
+        let _v = Vec::from_raw_parts(buf.ptr, buf.size, buf.size);
```

Comment:
This is incorrect: the third argument, capacity, might be wrong. I think you either need to add `cap` field to `Buffer`, or change `release_vec` to call `shrink_to_fit`. I'd also adive to use `Vec::into_raw` instead of `mem::forget` inside `release_vec`. If you use both `shrint_to_fit` and `into_raw`, it makes sence to add an assert that capacigty and length are indeed equal

Reply:
I assume that we pass a pre-shrunk vector (heap allocated slice) when creating the Buffer in release_vec. This should only be called by the external code with a Buffer previously returned by release_vec. In fact if it is used any other way, I am almost certain there will be great errors (is there a way to enforce that?)

However, I see the point of making both size and capacity explicit when passing these Buffers across FFI and I will update that. I think this will make the whole memory issue clearer.

Thanks for the reflection

Reply:
I think just adding an `assert!(v.len() == v.capacity())` to `release_vec` would be fine, if it's contract is that the buffer should b pre-shrunk. OTOH, `.shrink_to_fit` is an indempotent (and cheap in the shrunk case) operation, so one might as well just call it instead of an assert

Reply:
I cannot find `Vec::into_raw()` anywhere here: 
https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html

I would love to use it actually

Reply:
Huh, it looks like I've just imagined that this function exists, sorry for the noise! :unamused: